MALI calving dt convergence tests

matthewhoffman opened the pull request:
This merge adds a new set of tests for MALI that explore convergence with timestep of the calving physics.  This is added in a new test group that runs a configuration multiple times with the adaptive timestepper restricted by different fractions of the calculated calving CFL limit.  The results are then analyzed and a summary plot is produced.  This test can be applied to a number of different calving laws and meshes and using either data velocity or prognostic velocity.  These set of tests are not intended to be run routinely as part of integration, but instead are a tool for evaluating development of calving physics.  A test suite is also included that runs a number of these tests that have been evaluated as being useful.  There are some known ways this test group could be improved, but given its limited intended usage, they are not worth developing further at this time.

A MISMIP+ smoke test is also added as part of this PR.  The calving tests were first developed to use the MISMIP+ configuration, which was not yet in COMPASS.  Later, things were reorganized to make the calving tests more general and support multiple meshes.  Having a MISMIP+ smoke test is generally useful, so that test was retained in this branch after reorganizing things.

Note that this PR includes an update to the MALI-Dev submodule to include the merged MALI PR here: https://github.com/MALI-Dev/E3SM/pull/33.

pep8speaks commented:
Hello @matthewhoffman! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`compass/landice/tests/calving_dt_convergence/__init__.py`](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/__init__.py):

> [Line 33:5](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/__init__.py#L33): [E129](https://duckduckgo.com/?q=pep8%20E129) visually indented line with same indent as next logical line
> [Line 34:33](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/__init__.py#L34): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

* In the file [`compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py`](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py):

> [Line 9:1](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L9): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 32:1](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L32): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 50:5](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L50): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 54:5](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L54): [E301](https://duckduckgo.com/?q=pep8%20E301) expected 1 blank line, found 0
> [Line 59:9](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L59): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 60:9](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L60): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 65:46](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L65): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 95:13](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L95): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 103:9](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py#L103): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

* In the file [`compass/landice/tests/calving_dt_convergence/run_model.py`](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/run_model.py):

> [Line 4:1](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/run_model.py#L4): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 19:80](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/run_model.py#L19): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 87:80](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/run_model.py#L87): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 102:73](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/run_model.py#L102): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 102:80](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/run_model.py#L102): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 115:9](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/calving_dt_convergence/run_model.py#L115): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)





* In the file [`compass/landice/tests/mismipplus/smoke_test.py`](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/mismipplus/smoke_test.py):

> [Line 5:1](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/mismipplus/smoke_test.py#L5): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 31:5](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/mismipplus/smoke_test.py#L31): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 35:5](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/mismipplus/smoke_test.py#L35): [E301](https://duckduckgo.com/?q=pep8%20E301) expected 1 blank line, found 0
> [Line 46:1](https://github.com/MPAS-Dev/compass/blob/2e4f16097d880cfba779b187027ac633bd51f758/compass/landice/tests/mismipplus/smoke_test.py#L46): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

matthewhoffman commented:
@trhille , this is ready for review from you.  I haven't added the docs yet, but the code changes are done and I thought you might want to start looking at this before I have the docs complete.  To test this, you'll need to be using the accompanying MALI PR: https://github.com/MALI-Dev/E3SM/pull/33.  I'll update the submodule in this PR after we merge that.   For testing, I recommend running the new `calving_dt_convergence` test suite.  Note that the last two tests in it use the FO solver and take a few hours each.  But the non-FO tests collectively should finish in about an hour, so I think it would be acceptable to just test those (e.g. if your job times out before the FO tests finish).  When the tests are complete, just have a look at the png image that gets generated in each test directory.  The results should look similar to the images I've compiled here: https://docs.google.com/presentation/d/1YohEQBxoV84qVwrUJVM-WoB51oOv4eqTblqGx6ML3SE/edit?usp=sharing

trhille commented on compass/landice/tests/calving_dt_convergence/__init__.py:
```diff
@@ -0,0 +1,38 @@
+                            ]:
+                    if (calving == 'specified_calving_velocity' and
+                       velo == 'FO'):
+                        continue  # This combination is not useful
```
Why is this not useful?

matthewhoffman replied:
I was just trying to minimize adding too many new test cases, especially the FO ones, which are quite slow to run.  We can add them if you think it's worthwhile.

trhille replied:
No, I convinced myself that it's fine as-is, since we care about calving velocity and not retreat velocity for this CFL condition.

trhille commented on compass/landice/tests/calving_dt_convergence/albany_input.yaml:
```diff
@@ -0,0 +1,265 @@
+%YAML 1.1
```
I don't understand where this file is used. I see that the Humboldt, Thwaites, and MISMIP+ cases all use albany_input.yaml files that are in their respective test group directories.

trhille commented on compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py:
```diff
@@ -0,0 +1,109 @@
+
+        Parameters
+        ----------
+        test_group : compass.landice.tests.mismipplus.MISMIPplus
```
This doesn't belong to the mismipplus test group, right?

trhille commented on compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py:
```diff
@@ -0,0 +1,109 @@
+        Test cases can override this method to perform validation of variables
+        and timers
+        """
+        # variables = ['thickness', 'surfaceSpeed']
```
Maybe it would be good to add a note here that if you want to use compare_variables, you should uncomment line 3 as well

trhille commented on compass/landice/tests/calving_dt_convergence/namelist.landice:
```diff
@@ -0,0 +1,30 @@
+config_tracer_advection = 'none'
+config_thermal_solver = 'none'
+config_temperature_init = 'file'
+config_dynamic_thickness = 0.001
```
Why this value?

trhille commented on compass/landice/tests/calving_dt_convergence/namelist.landice:
```diff
@@ -0,0 +1,30 @@
+config_start_time = '0000-01-01_00:00:00'
+config_run_duration = '0025-00-00_00:00:00'
+config_adaptive_timestep = .true.
+config_adaptive_timestep_CFL_fraction = 0.95
```
Why this value instead of the usual 0.8?

matthewhoffman replied:
To be as aggressive as possible with preventing advection from limiting the CFL.

trhille commented on compass/landice/tests/calving_dt_convergence/run_model.py:
```diff
@@ -0,0 +1,133 @@
+
+class RunModel(Step):
+    """
+    A step for performing forward MALI runs as part of MISMIP+ test cases.
```
dt convergence test case

trhille commented on compass/landice/tests/mismipplus/albany_input.yaml:
```diff
@@ -0,0 +1,265 @@
+
+# Discretization Description
+  Discretization:
+    Exodus Output File Name: albany_output.exo
```
Could comment this out, unless we need these exo files for any reason.

trhille commented on compass/landice/tests/mismipplus/namelist.landice:
```diff
@@ -0,0 +1,27 @@
+config_start_time = '0000-01-01_00:00:00'
+config_run_duration = '0005-00-00_00:00:00'
+config_adaptive_timestep = .true.
+config_adaptive_timestep_CFL_fraction = 0.9
```
Why is this different from the CFL fraction of 0.95 in https://github.com/MPAS-Dev/compass/pull/386/files#diff-acd02e538bc2c87b44f42baa553a3630dd419f20c83568eac48f8b2a9fc497b7 ?

trhille replied:
(that link is to the namelist.landice file in dt_convergence_test)

trhille reviewed (changes requested):
@matthewhoffman, this is really impressive! I've finished going through the code (only found minor copy-paste errors and had a few questions), but I'm still running the tests.

trhille commented:
I ran the `calving_dt_convergence` test suite on Badger using an exe compiled at the head of `matthewhoffman/MALI/calving_CFL` (commit 258eadc1c90a616a26675538088dad383d726ed3). 
`humboldt.specified_calving_velocity.none`:
![image](https://user-images.githubusercontent.com/17446278/167674560-3607d9b4-73a1-49b4-91f4-66daff6c7c72.png)

`humboldt.von_Mises_stress.FO`
![image](https://user-images.githubusercontent.com/17446278/167674726-01241b81-9754-40c1-b831-7fe87a9656cc.png)

`humboldt.von_Mises_stress.none`
![image](https://user-images.githubusercontent.com/17446278/167675052-e59ec6e6-4136-4f33-85f4-c6f7b3aec3bc.png)

`mismip+.specified_calving_velocity.none`
![image](https://user-images.githubusercontent.com/17446278/167675123-21d520a5-8a08-4ffe-b986-1758601b765f.png)

`mismip+.von_Mises_stress.FO`
![image](https://user-images.githubusercontent.com/17446278/167675287-3083421a-234e-46a9-b56e-1f34df6a8813.png)

`mismip+.von_Mises_stress.none`
![image](https://user-images.githubusercontent.com/17446278/167675401-9b12024c-13e4-44c8-aeff-4493c107b99e.png)

`thwaites.specified_calving_velocity.none`
![image](https://user-images.githubusercontent.com/17446278/167675498-a6ed4642-1bf6-47b2-b2ec-0c67eb061a00.png)

`thwaites.von_Mises_stress.FO`
![image](https://user-images.githubusercontent.com/17446278/167675625-5279ac26-a4ac-46b2-8fa0-c745f53c6c73.png)

`thwaites.von_Mises_stress.none`
![image](https://user-images.githubusercontent.com/17446278/167675745-83850cd4-d8c0-4168-a9ce-268642f50444.png)

trhille commented on compass/landice/tests/calving_dt_convergence/dt_convergence_test/__init__.py:
```diff
@@ -0,0 +1,109 @@
+            i += 1
+
+        ax[0].legend(loc='best', prop={'size': 5})
+        plt.savefig('calving_dt_comparison.png', dpi=150)
```
Can we give each .png file the same name as the test case? It would be nice to be able to scp all the figures into one directory for perusal without having to rename them individually.

trhille reviewed (changes requested):
One more change I thought of while looking through output figures.

matthewhoffman commented:
@trhille , thanks for reviewing this and checking all the output images.  I've pushed a commit addressing your review comments.  I'm going to work on the docs next, but feel free to look it over now or wait until I have the docs ready.

matthewhoffman commented:
@xylar , this landice PR is finished and just awaiting CI to complete and a final re-run of our integration suite.  If you want to review it before Trevor's final re-review and merge, let us know.

matthewhoffman commented:
Update: CI and test suites all pass/give expected behavior

xylar commented:
Nope, go ahead.  It looks like there are some conflicts.  As you know, my preference would be to rebase to handle them but you could either merge `master` to this branch or fix them in a merge commit if you prefer.

trhille commented:
All tests passed upon re-running the `calving_dt_convergence` test suite on Badger with an exe compiled on the head of MALI-Dev/develop (commit 9d3dcd9b0a1a15e20458a7e3394b545517ae45b5)
```
Test Runtimes:
02:40 ESC[92mPASSESC[0m landice_calving_dt_convergence_mismip+.specified_calving_velocity.none
02:08 ESC[92mPASSESC[0m landice_calving_dt_convergence_mismip+.von_Mises_stress.none
01:56 ESC[92mPASSESC[0m landice_calving_dt_convergence_humboldt.specified_calving_velocity.none
01:39 ESC[92mPASSESC[0m landice_calving_dt_convergence_humboldt.von_Mises_stress.none
02:49 ESC[92mPASSESC[0m landice_calving_dt_convergence_thwaites.specified_calving_velocity.none
03:16 ESC[92mPASSESC[0m landice_calving_dt_convergence_thwaites.von_Mises_stress.none
72:48 ESC[92mPASSESC[0m landice_calving_dt_convergence_mismip+.von_Mises_stress.FO
08:35 ESC[92mPASSESC[0m landice_calving_dt_convergence_humboldt.von_Mises_stress.FO
21:20 ESC[92mPASSESC[0m landice_calving_dt_convergence_thwaites.von_Mises_stress.FO
Total runtime 117:12
PASS: All passed successfully!
```

trhille commented:
Figures look very similar to previous, but not totally identical. This could be due to the small changes in namelist options introduced after my last review, or maybe to the small cleanup from the review of MALI-Dev PR 33 (less likely). Unfortunately I forgot to uncomment the variable comparison lines with I ran, so the baseline comparison did not work. Just comparing these visually:
`humboldt_specified_calving_velocity_none` (top few curves in third panel are subtly but noticeably different from before):
![Uploading calving_dt_convergence_test_humboldt_specified_calving_velocity_none_comparison_plot.png…]()

The other test cases are not noticeably different than before.

trhille commented on docs/developers_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -0,0 +1,41 @@
+
+There are no shared config options for the ``mismipplus`` test group.
+
+The mismipplus test can only be run with the FO velocity solver.
```
Can't it also be run with velocity solver off?

trhille commented on docs/developers_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -0,0 +1,41 @@
+~~~~~~~~~
+
+The class :py:class:`compass.landice.tests.mismipplus.run_model.RunModel`
+defines a step for running MALI an initial condition downloaded from
```
Missing "using" or "from" after "running MALI"

trhille commented on docs/developers_guide/landice/test_groups/mismipplus.rst:
```diff
@@ -0,0 +1,41 @@
+----------
+
+The :py:class:`compass.landice.tests.mismipplus.smoke_test.SmokeTest`
+performs a 5 year run.  There is a validation step that compares th output
```
typo in "th"

trhille commented on docs/users_guide/landice/suites.rst:
```diff
@@ -5,7 +5,8 @@ Test suites
 
 The ``landice`` core currently includes 5 :ref:`test_suites` that can be used
```
might be 6 suites now, depending on whether full_integration counts as its own.

trhille commented on docs/users_guide/landice/suites.rst:
```diff
@@ -142,3 +147,29 @@ The suite includes:
+calving physics for a number of different model meshes, calving laws, and
+velocity solver options.  The tests with "none" velocity solver use data
+velocity fiels, and each take 5-10 minutes.  The tests with FO velocity
+solver each take a few hours and one may prefer to run them in individual
```
These only took ~100 minutes total for me on Badger:
```
72:48 ESC[92mPASSESC[0m landice_calving_dt_convergence_mismip+.von_Mises_stress.FO
08:35 ESC[92mPASSESC[0m landice_calving_dt_convergence_humboldt.von_Mises_stress.FO
21:20 ESC[92mPASSESC[0m landice_calving_dt_convergence_thwaites.von_Mises_stress.FO
```

trhille commented on docs/users_guide/landice/suites.rst:
```diff
@@ -142,3 +147,29 @@ The suite includes:
+The ``calving_dt_convergence`` test suite runs timestep convergence tests for
+calving physics for a number of different model meshes, calving laws, and
+velocity solver options.  The tests with "none" velocity solver use data
+velocity fiels, and each take 5-10 minutes.  The tests with FO velocity
```
These took ~15 minutes total for me on Badger:
```
02:40 ESC[92mPASSESC[0m landice_calving_dt_convergence_mismip+.specified_calving_velocity.none
02:08 ESC[92mPASSESC[0m landice_calving_dt_convergence_mismip+.von_Mises_stress.none
01:56 ESC[92mPASSESC[0m landice_calving_dt_convergence_humboldt.specified_calving_velocity.none
01:39 ESC[92mPASSESC[0m landice_calving_dt_convergence_humboldt.von_Mises_stress.none
02:49 ESC[92mPASSESC[0m landice_calving_dt_convergence_thwaites.specified_calving_velocity.none
03:16 ESC[92mPASSESC[0m landice_calving_dt_convergence_thwaites.von_Mises_stress.none
```

trhille reviewed (approved):
@matthewhoffman, great work on this! I've noted a few totally nit-picky things, but am also approving it since these are trivial and do not affect functionality.
